mdn/developer-portal
# 945: Update pagination and filtering to be done on the server side, not the client side

This changeset addresses the problem posed by the current implementation of clientside psuedo paging/filtering: because the page loads all the possible results and then shows/hides DOM elements to match, the overall page weight is already getting significant (eg /posts/ is 10MB and pulls in 18MB of images)

So, when merged, this changeset will switch that to all being handled server-side. The trade-off is we do a full page load each time a page or filter parameter is changed, but if this is too much we can look to update this to AJAX calls + DOM-element replacement.

The original JS for the pagination and filtering has been repurposed and retained wherever possible to still benefit from work done to populate the filter form from querystring params and to show/hide "Clear" triggers appropriately.

## NOTES
- ~The UX of filtering has retained the original behaviour of updating result _immediately_ after ticking a filter box. This now triggers a page reload, which is ok/tolerable on desktop, but more annoying on mobile, because the Filters controls are collapsed when the page is loaded again.~
~This could be dealt with either:~
  I WENT WITH THIS ONE ~1. By changing the filtering to have a 'Filter results' button as the trigger - improving both Mobile and Desktop UX, potentially~
  ~2. On mobile, ensure that if a filter is selected, the Filters panel is kept open on load. This isn't ideal, but would help a little.~
~(My preference is for 1 -- @atopal, what do you think?)~

- DONE ~On Events, future events are not paginated, while past events are. We may well move Past events to their own page, but for now, because past events are so far down the page, viewing the next page is a click-load-scroll down situation.~ 
~This could be dealt with by:~
  * ~Slotting in an #anchor in the pagination links so that the user is auto-scrolled down the appropriate part of the events page after load -- @atopal, worth it?~ DONE

(Resolves #945)

## How to test

TO COME: I'm going to get this branch deployed to a new development environment so that stakeholders can try it out without us having to merge the branch to `master` and we can do direct comparisons in performance and page size with current production. 

If this can't be done soonish, I can also expose by local build over an SSH tunnel - just DM me in Slack and I'll enable it. It won't be as good a comparison, but it'll allow manual test drives.

Note that when there are multiple 'facets' to filtering (eg topic and location) these are combined as `AND` but _within_ each facet, the selections are X `OR` Y.

Areas to test:
- `/posts/`  - try the filtering and pagination. 
- `/events/`  - try the filtering for upcoming events and pagination for past events. Note that the past events listing is not affected by the pagination  
- `/communities/people/` - try filtering and pagination

# Screenshots
![Screenshot_2019-12-13 Events](https://user-images.githubusercontent.com/101457/70811018-b8d14100-1dbc-11ea-9319-ed1e55527aca.png)
![Screenshot_2019-12-13 People](https://user-images.githubusercontent.com/101457/70811019-b8d14100-1dbc-11ea-8d93-9284276ef256.jpg)
![Screenshot_2019-12-13 People](https://user-images.githubusercontent.com/101457/70811020-b8d14100-1dbc-11ea-83cf-8dbbdf64c25c.png)
![Screenshot_2019-12-13 People(1)](https://user-images.githubusercontent.com/101457/70811021-b8d14100-1dbc-11ea-9570-6cc53ebf14f3.png)
![Screenshot_2019-12-13 Posts](https://user-images.githubusercontent.com/101457/70811022-b8d14100-1dbc-11ea-9e7a-f9983e177deb.png)


<img width="342" alt="Screenshot 2019-12-13 at 12 44 33" src="https://user-images.githubusercontent.com/101457/70810979-ac4ce880-1dbc-11ea-8cc4-327695ae8774.png">
